### PR TITLE
[pipeline]  Llama pipeline

### LINKED ISSUE #4204: [shardformer] support pipeline for llama
add llama forward and policy

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
closes #4204

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This pr adds llama model pipeline and tests. The forward method receives a stage index list to get the start and end layer

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/policies/llama.py [hunk 1]
```diff
@@ -144,3 +194,151 @@ def module_policy(self):
...
+    position_ids: Optional[torch.LongTensor] = None,
+    past_key_values: Optional[List[torch.FloatTensor]] = None,
+    inputs_embeds: Optional[torch.FloatTensor] = None,
+    use_cache: Optional[bool] = None,
+    output_attentions: Optional[bool] = None,
+    output_hidden_states: Optional[bool] = None,
+    return_dict: Optional[bool] = None,
+    stage_manager: Optional[PipelineStageManager] = None,
+    hidden_states: Optional[torch.FloatTensor] = None,
+    stage_index: Optional[List[int]] = None,
```
Comment (ver217) [inline colossalai/shardformer/policies/llama.py:212]: Where do you bind this argument?

File: tests/test_shardformer/test_model/test_shard_llama_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    pg_mesh = ProcessGroupMesh(DP_SIZE, PP_SIZE)
+    stage_manager = PipelineStageManager(pg_mesh, PP_DIM)
+
+    sub_model_zoo = model_zoo.get_sub_registry('transformers_llama')
+    x = torch.randint(0, 1000, (2, 3)).cuda()
+    hidden_states = torch.randint(0, 1000, (2, 3, 128)).to(torch.float32).cuda()
+    for name, (model_fn, data_gen_fn, output_transform_fn, loss_fn, _) in sub_model_zoo.items():
+        if name == 'transformers_llama':
+            org_model, sharded_model = build_pipeline_model(model_fn, stage_manager, enable_fused_normalization,
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_shard_llama_pipeline.py:53]: You should use shardformer here.

File: colossalai/shardformer/policies/llama.py [hunk 3]
```diff
@@ -144,3 +194,151 @@ def module_policy(self):
...
+                                    seq_length + past_key_values_length,
+                                    dtype=torch.long,
+                                    device=device)
+        position_ids = position_ids.unsqueeze(0).view(-1, seq_length)
+    else:
+        position_ids = position_ids.view(-1, seq_length).long()
+
+    # embed positions, for the first stage, hidden_states is the input embeddings,
+    # for the other stages, hidden_states is the output of the previous stage
+    # TODO: we should recive the attn mask of 1st stage and send it to the other stages
```
Comment (ver217) [inline colossalai/shardformer/policies/llama.py:269]: Is this comment outdated?

File: colossalai/shardformer/policies/llama.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/llama.py:216]: done

File: tests/test_shardformer/test_model/test_shard_llama_pipeline.py (hunk 2 above)
Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_shard_llama_pipeline.py:53]: the `build_pipeline_model` is calling it

File: colossalai/shardformer/policies/llama.py (hunk 3 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/llama.py]: done

File: tests/test_shardformer/test_model/test_shard_llama_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                                                            enable_tensor_parallelism, use_lazy_init)
+            if stage_manager.stage == 0:
+                attention_mask = torch.ones_like(x).cuda()
+                output = sharded_model(input_ids=x, attention_mask=attention_mask, stage_manager=stage_manager)
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_shard_llama_pipeline.py:57]: If you replace forward correctly, I think you don't need to pass stage_manager here.

Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_shard_llama_pipeline.py]: done

**Review by ver217 — APPROVED** [review]

**Comment by CjhHa1** [comment]
截屏2023-07-11 11 36 15

Rerun failed test

**Pull request merged**